qiskit/aqua/operators/list_ops/composed_op.py

```diff
@@ -62,6 +63,13 @@ def distributive(self) -> bool:
     #     """ Tensor product with Self Multiple Times """
     #     raise NotImplementedError
 
+    def to_circuit(self) -> QuantumCircuit:
+        """ Returns the quantum circuit, representing the composed operator.
+        Returns:
+            The circuit representation of the composed operator.
+        """
+        return self.to_circuit_op().to_circuit()  # type: ignore
```

Comment:
Why is there a `type: ignore` necessary? Shouldn't mypy know that `to_circuit` always returns a `QuantumCircuit`? (CircuitOp should have this return type as well)

Reply:
because to_circuit_op() returns 'OperatorBase'

Reply:
maybe we can change the return type of to_circuit_op() to Union[CircuitOp, CircuitStateFn] ?

Reply:
Although you moved this whole method from ListOp down to here it still relies the to_circuit_op of ListOp which can, dependent on the overridden implementation of reduce end up either with a list of operators or a single one. Hence in terms of code I do not see any change in behavior and the comments I made originally about lists still stand. The test worked before (I did not look in detail at the test again) because in that instance it was a single PrimitiveOp being returned by reduce (as called in to_circuit_op) not a list. This list versus single needs to be dealt with. Maybe a better to_circuit_op in the derived class that gives a single operator for the entity simplifying it, or whatever to achieve this. Maybe this should give back a (nested) list of circuits - one for every circuitop that comes back from to_circuit_op. Or we just support the simple case where it does come back with a single circuit_op promitive - which I think is really all this code supports and raises an exception otherwise. Clearly a more general solution is preferable but seemingly more complex

Reply:
Both works, yes, maybe for the time being just handling the case of a single circuit works. 👍

Reply:
I took closer look on reduce operations and if we don’t include EvolvedOps, it works as follows:
1) if we have (primitive_op_1 ^ primitive_op_2), then tensored_op.reduce returns one CircuitOp (which is PrimitiveOp)
2) if we have (primitive_op_1 @ primitive_op_2), then composed_op.reduce returns PrimitiveOp

Thus if we have expression using only MatrixOps, PauliOps, CircuitOps, @ and ^,  to_circuit_op will recursively apply 1) and 2) and we will end up with one CircuitOp with one underlying circuit.

Reply:
Ok, so lets go with the single circuit return then so if the result of to_circuit_op() is not an instance of ListOp then call to_circuit() on that and return the QuantumCircuit produced else raise an error to the effect that the conversion is only supported where a single circuit is produced from the operator.

Reply:
Ok, I just addressed it and to avoid #type: ignore, I check if the instance is PrimitiveOp or CircuitStateFn